sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -82,36 +84,181 @@ private DoFnSignatures() {}
     return signature;
   }
 
+  /**
+   * The global context for the class, for use in analysis.
+   *
+   * <p>It contains much of the information that eventually becomes part of the {@link
+   * DoFnSignature}, but in an intermediate state.
+   */
+  @AutoValue
+  abstract static class GlobalAnalysisContext {
```

Comment:
One of the more annoying bits is that this mirrors `DoFnSignature` quite a lot. But I want the public API for `DoFnSignature` to not be full of `@Nullable` fields that a user has to be careful about. I think that actually expansion of this probably obsoletes `DoFnSignature.Builder`.

If you have any other clever ideas, I'm open to them.